onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -49,487 +76,553 @@ struct Provider_IDeviceAllocator_Impl : Provider_IDeviceAllocator {
   std::unique_ptr<IDeviceAllocator> p_;
 };
 
-struct Provider_TensorProto_Impl : ONNX_NAMESPACE::Provider_TensorProto {
-  Provider_TensorProto_Impl(ONNX_NAMESPACE::TensorProto* p) : p_{p} {}
+struct Provider_TensorShapeProto_Dimension_Iterator_Impl : Provider_TensorShapeProto_Dimension_Iterator {
+  Provider_TensorShapeProto_Dimension_Iterator_Impl(google::protobuf::internal::RepeatedPtrIterator<const onnx::TensorShapeProto_Dimension>&& v) : v_{std::move(v)} {}
 
-  void CopyFrom(const Provider_TensorProto& v) override {
-    *p_ = *static_cast<const Provider_TensorProto_Impl*>(&v)->p_;
-  }
+  bool operator!=(const Provider_TensorShapeProto_Dimension_Iterator& p) const override { return v_ != static_cast<const Provider_TensorShapeProto_Dimension_Iterator_Impl*>(&p)->v_; }
 
-  ONNX_NAMESPACE::TensorProto* p_{};
+  void operator++() override { v_.operator++(); }
+  const Provider_TensorShapeProto_Dimension& operator*() override { return *reinterpret_cast<const Provider_TensorShapeProto_Dimension*>(&v_.operator*()); }
+
+  google::protobuf::internal::RepeatedPtrIterator<const onnx::TensorShapeProto_Dimension> v_;
 };
 
-struct Provider_AttributeProto_Impl : ONNX_NAMESPACE::Provider_AttributeProto {
-  Provider_AttributeProto_Impl() = default;
-  Provider_AttributeProto_Impl(const ONNX_NAMESPACE::AttributeProto& copy) : v_{copy} {}
+struct Provider_NodeAttributes_Iterator_Impl : Provider_NodeAttributes_Iterator {
+  Provider_NodeAttributes_Iterator_Impl(NodeAttributes::const_iterator&& v) : v_{std::move(v)} {}
 
-  std::unique_ptr<Provider_AttributeProto> Clone() const override {
-    return onnxruntime::make_unique<Provider_AttributeProto_Impl>(v_);
-  }
+  bool operator!=(const Provider_NodeAttributes_Iterator& p) const override { return v_ != static_cast<const Provider_NodeAttributes_Iterator_Impl*>(&p)->v_; }
 
-  ::onnx::AttributeProto_AttributeType type() const override { return v_.type(); }
+  void operator++() override { v_.operator++(); }
+  const std::string& first() const override { return v_->first; }
+  const Provider_AttributeProto& second() override { return *reinterpret_cast<const Provider_AttributeProto*>(static_cast<const ONNX_NAMESPACE::AttributeProto*>(&v_->second)); }
```

Comment:
>reinterpret_cast<const Provider_AttributeProto* [](start = 61, length = 47)

why not cast straight to const reference?

Reply:
I just always used pointer reinterpret_casts since I've had unexpected weird issues with references sometimes. This way just felt safer and more explicit.